src/services/refactors/moveToNewFile.ts

```diff
@@ -0,0 +1,359 @@
+        const okStart = range.pos <= statements[startNodeIndex].getStart(file);
+        const afterEndNodeIndex = findIndex(statements, s => s.end > range.end, startNodeIndex);
+        // Can't be partially into the next node
+        const okEnd = afterEndNodeIndex === -1 || afterEndNodeIndex !== 0 && statements[afterEndNodeIndex].getStart(file) >= range.end;
+        return okStart && okEnd ? statements.slice(startNodeIndex, afterEndNodeIndex === -1 ? statements.length : afterEndNodeIndex) : undefined;
+    }
+
+    function doChange(oldFile: SourceFile, program: Program, toMove: ReadonlyArray<Statement>, changes: textChanges.ChangeTracker, host: LanguageServiceHost): void {
+        const checker = program.getTypeChecker();
+        const usage = getUsageInfo(oldFile, toMove, checker);
+
+        const currentDirectory = getDirectoryPath(oldFile.fileName);
+        const extension = extensionFromPath(oldFile.fileName);
+        const newModuleName = makeUniqueModuleName(getNewModuleName(usage.movedSymbols), extension, currentDirectory, host);
+        const newFileNameWithExtension = newModuleName + extension;
+
+        // If previous file was global, this is easy.
+        changes.createNewFile(oldFile, combinePaths(currentDirectory, newFileNameWithExtension), getNewStatements(oldFile, usage, changes, toMove, program, newModuleName));
+    }
+
+    function getNewStatements(
+        oldFile: SourceFile, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ReadonlyArray<Statement>, program: Program, newModuleName: string,
+    ): ReadonlyArray<Statement> {
+        const checker = program.getTypeChecker();
+
+        if (!oldFile.externalModuleIndicator) {
+            changes.deleteNodeRange(oldFile, first(toMove), last(toMove));
+            return toMove;
+        }
+
+        const importsFromNewFile = createOldFileImportsFromNewFile(usage.oldFileImportsFromNewFile, newModuleName);
+        if (importsFromNewFile) {
+            changes.insertNodeBefore(oldFile, oldFile.statements[0], importsFromNewFile, /*blankLineBetween*/ true);
+        }
+
+        deleteUnusedOldImports(oldFile, toMove, changes, usage.unusedImportsFromOldFile, checker);
+        changes.deleteNodeRange(oldFile, first(toMove), last(toMove));
+
+        updateImportsInOtherFiles(changes, program, oldFile, usage.movedSymbols, newModuleName);
+
+        return [
+            ...getNewFileImportsAndAddExportInOldFile(oldFile, usage.oldImportsNeededByNewFile, usage.newFileImportsFromOldFile, changes, checker),
+            ...addExports(toMove, usage.oldFileImportsFromNewFile),
+        ];
+    }
+
+    function deleteUnusedOldImports(oldFile: SourceFile, toMove: ReadonlyArray<Statement>, changes: textChanges.ChangeTracker, toDelete: ReadonlySymbolSet, checker: TypeChecker) {
+        for (const statement of oldFile.statements) {
+            if (!contains(toMove, statement) && isImportDeclaration(statement)) {

```

Comment:
what about `ImportEqualsDeclaration`?

Reply:
also what should we do with require statements in a .js file. seems like we have enough information in the current file to know if we should create `import` or `const x = require(...)`..